### ISSUE TITLE
Multiple Metallic Mimics don't get counters

### ISSUE THREAD

**Issue by loetermann** [magefree/mage#2765]
If a Metallic Mimic is on the battlefield and a second Metallic Mimic enters, it should enter the battlefield with a +1/+1 counter, if the same creature type is chosen (as stated here http://magic.wizards.com/en/articles/archive/feature/aether-revolt-release-notes-2017-01-06 under Metallic Mimic).

In the current implementation it does not get a counter.

**Comment by loetermann** [comment]
I just looked into the code (for the first time, so I don't see the whole picture yet) and figured out that uncommenting [this line](https://github.com/magefree/mage/blob/master/Mage/src/main/java/mage/abilities/effects/ContinuousEffects.java#L884) fixes the problem but, I guess, creates other issues. For example I noticed that then (after uncommenting it), a Ghostly Flicker targeting two Mimics causes them to return with a counter each (if the same creature type as before flickering is chosen) or with a counter on one of them (if a different type is chosen).
Looking deeper in the code I noticed that this is due to the implementation of Ghostly Flicker (might need reworking?) as it puts the cards on the battlefield each by each instead of simultaneously.
Casting Displace on two Mimics works as expected (none of them getting a counter) even after uncommenting the above mentioned line. Are there concrete reasons for that line being commented out?

**Comment by mcockle23** [comment]
I've been debugging this for a while today.  I'm still pretty new to this codebase so bear with me.

First off, I tried uncommenting the call to game.applyEffects() in that spot did not fix the issue for me.

In fact, as far as I can tell, that call is what is causing this bug.  Every time anything happens, a land is played, a spell is cast, priority or phase changes, game.applyEffects() is called.  In that call, everything in the game is reset, and all effects are re-applied.  So, in this case, the second Mimic is cast, the original one gets reset to not be of the chosen type, and then it is re-applied.  Then, when the second one enters the battlefield, the first is reset to not be the chosen type, the check for whether or not it should get a +1/+1 happens, neither is of the chosen type.  Then they both get the effect applied to get their chosen type.

Like I said, I'm pretty new here, so I'm not exactly sure what the best course of action is, but that is what appears to be causing this to not work properly.

**Comment by loetermann** [comment]
I just tried it again and have to admit that I made a second change that is required.
In MetallicMimic.java I added the following lines after [line 98](https://github.com/magefree/mage/blob/450eba06e88d53365f8a6c832c8dcbeb5dab8731/Mage.Sets/src/mage/cards/m/MetallicMimic.java#L98)

    if (permanent == null) {
       permanent = game.getPermanentEntering(source.getSourceId());
    }

Without those lines, the mimic only applies the chosen type to itself, if it is a permanent on the battlefield. This way it also applies it as it is entering the battlefield.

Regarding your understanding of the cause of it, I have a different understanding. When calling game.applyEffects it resets the effects and re-applies them immediately. So the checking, whether the counter should be added, is done after re-applying and not in between. If I understand it correctly, the checking of the replacement effect is done in this [line](https://github.com/magefree/mage/blob/master/Mage/src/main/java/mage/abilities/effects/ContinuousEffects.java#L778) (not sure about it, though).

Anyway, my understanding of this was the following:

one mimic is on the battlefield
the second one is entering the battlefield

    ContinuousEffects.replaceEvent gets called
       while true {
         find applicable effect
         apply one effect or break
         // game.applyEffects
       }

So the first time the loop gets called, the "choose type" effect is applied. If we call game.applyEffects now, all effects would be reset (like you said) and **applied again**. At this point, both mimics become of the chosen type (the chosen one only with the modification mentioned above, that it also affects itself while entering). Then the loop starts again and finds the "add +1/+1 counter" effect and applies it.

It also works for me leaving ContinousEffects.java unchanged, if I modify the ChooseCreatureTypeEffect to 


        // As Metallic Mimic enters the battlefield, choose a creature type.
        this.addAbility(new AsEntersBattlefieldAbility(new ChooseCreatureTypeEffect(Outcome.BoostCreature) {
            @Override
            public boolean apply(Game game, Ability source) {
                boolean result = super.apply(game, source);
                game.applyEffects();
                return result;
            }
        }));

But as I mentioned, I'm new to the code as well, so I might be wrong with the order in which effects are reset and re-applied and checked...

My guessing what the problem with uncommenting the game.applyEffects line in ContinuousEffects is:
1. A continuous effect of a permanent **does** affect itself "while entering the battlefield".
2. Continuous Effects of other permanents (that are already on the battlefield) **don't affect** permanents "while those are entering".
3. Continuous Effects of a permanent **don't affect** other permanents "while this permanent is entering".

So if we don't call game.applyEffect none of those effects will be applied, so 2 and 3 are correct, 1 is not working correctly, like in case of the mimic. If we do game.applyEffect while permanents are entering the battlefield, every continuous effect has to check that it handles the different cases correctly. Basically 2 will be working correctly with every effect that just looks for permanents on the battlefield and not those that are entering. And 3 will be correct, if each continuous effect checks whether it's source is a permanent.
Furthermore it will cause problems with effects that put permanents on the battlefield simultaneously but are implemented incorrectly, as the continuous effect of the first one that entered will be applied to the following ones. This was the problem with GhostlyFlicker (already fixed).

**Comment by loetermann** [comment]
I don't understand all of your changes, but I think you replaced Metallic Mimics first static ability "Metallic Mimic is the chosen type in addition to its other types." by an OneShotEffect. I didn't test it, but I think this will cause other problems. For example a Humble ("Until end of turn, target creature loses all abilities and has base power and toughness 0/1.") should reset its types. Not sure about it though.

**Comment by mcockle23** [comment]
Yeah my analysis was off due to the nature of my test. I went back and noticed that the mimic i saw as not having the selected type was the computer's and it's because it chose shapeshifter as the type, so it looked like one of mine without the selected type.

This was caused by the entering mimic not having the selected type as it enters the battlefield, it was adding the type after it entered and after the check for the first one to give it a counter. So the fix works by giving it the type as it enters. 

Well done! I played a few matches with it and it's working well now.
